assets/src/edit-story/utils/usePageAsCanvas.js

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { useCallback, useMemo } from 'react';
+import * as htmlToImage from 'html-to-image';
+import { FULLBLEED_RATIO, getBox, PAGE_RATIO } from '@web-stories-wp/units';
+
+/**
+ * Internal dependencies
+ */
+import useLibrary from '../components/library/useLibrary';
+import { useCanvas, useLayout, useStory } from '../app';
+import { ZOOM_SETTING } from '../constants';
+import { getAccessibleTextColorsFromPixels } from './contrastUtils';
+import { calculateTextHeight } from './textMeasurements';
+
+function usePageAsCanvas() {
+  const { pageCanvasData, setPageCanvasData } = useLibrary((state) => ({
+    pageCanvasData: state.state.pageCanvasData,
+    setPageCanvasData: state.actions.setPageCanvasData,
+  }));
+  const { fullbleedContainer } = useCanvas((state) => ({
+    fullbleedContainer: state.state.fullbleedContainer,
+  }));
+  const { currentPage } = useStory(({ state }) => {
+    return {
+      currentPage: state.currentPage,
+    };
+  });
+  const { zoomSetting, setZoomSetting } = useLayout(
+    ({ state: { zoomSetting }, actions: { setZoomSetting } }) => ({
+      zoomSetting,
+      setZoomSetting,
+    })
+  );
+
+  const pageHash = useMemo(() => {
+    return JSON.stringify(currentPage);
+  }, [currentPage]);
+
+  const hasPageHashChanged = useCallback(() => {
+    return pageHash !== pageCanvasData.currentPage;
+  }, [pageHash, pageCanvasData]);
+
+  const generateCanvasFromPage = useCallback(
+    (callback, resetZoom = false) => {
+      // If we're not resetting the zoom and we're zoomed in, skip pre-generation of the canvas.
+      // We'll be resetting the zoom later when inserting any element so it would be not useful.
+      if (!resetZoom && zoomSetting !== ZOOM_SETTING.FIT) {
+        return;
+      }
+      if (
+        !pageCanvasData ||
+        hasPageHashChanged(currentPage, pageCanvasData.currentPage)
+      ) {
+        if (resetZoom) {
+          // If we're inserting an element, we need to reset zoom, too.
+          setZoomSetting(ZOOM_SETTING.FIT);
+        }
+        // Wait one tick for the zoom to settle in.
+        setTimeout(() => {
+          htmlToImage
+            .toCanvas(fullbleedContainer, {
+              fontEmbedCss: '',
+            })
+            .then((canvas) => {
+              if (callback) {
+                callback(canvas);
+              } else {
+                setPageCanvasData({
+                  canvas,
+                  currentPage: pageHash,
+                });
+              }
+            })
+            .catch(() => resetZoom && callback());
+        });
+      }
+    },
+    [
+      currentPage,
+      fullbleedContainer,
+      pageCanvasData,
+      setPageCanvasData,
+      setZoomSetting,
+      zoomSetting,
+      hasPageHashChanged,
+      pageHash,
+    ]
+  );
+
+  const calculateAccessibleTextColors = useCallback(
```

Comment:
Do we have a rough idea on how much latency this introduces? Both canvas generation and color calculation. If significant, maybe we can consider doing this async e.g. on background changes.

Since we're about to head into a performance sprint anyways. :)

Reply:
The canvas generation latency is significant (1+ seconds last time I tested, IIRC).
Currently, the canvas generation is started already when hovering over the text button or in the area where text presets are -- so it'll be ready in most cases once the user clicks to actually add the preset, and only the added color calculation time will be relevant. Only if the canvas isn't already ready will it be generated at the same time with clicking.

If we'd re-generate the canvas every time there's a change on the page, it would probably cause general performance issues -- unfortunately we can't only check the background -- shapes, other text elements etc. all influence the color.

Color calculation took less time (below 200ms always when I tested) -- but I'll check again tomorrow, there have been some changes since I last tested that.

Reply:
I was looking into doing that in the background, but the first and main piece - canvas image generation cannot be done in Web Worker, and without that, I think it will introduce a general slowdown like Miina said (as opposed to waiting 1+ seconds when this feature is used).

Reply:
@choumx @merapi  I've added [an additional optimisation](3b08371) to ensure that the canvas generation is not happening multiple times and if there's an unresolved Promise, the same is used.

From testing:
- It looks like the image generation is taking less time than I measured before, currently it averaged around 500ms.
- The color calculation takes few time, around 10-30ms.

It looks like there are other things lagging as well, e.g. even if the image is pre-generated then there's still a lag when inserting a text to the page -- especially if there are many elements on the page. It seems like re-rendering the page is adding a visible lag. Perhaps some performance analysis of the initial editor render would be a good task for the performance sprint.

Reply:
Can we feature flag this change? 

Given existing performance issues, I think we may need to disable this feature for slower devices e.g. perhaps some max latency on canvas generation + color calculation that toggles a user setting for this feature.

Reply:
Ok, I'll add it behind a feature flag for now. This needs some more discussion then about if and when to enable this for the users. cc @barklund for information

Reply:
Done in d8b127f

Reply:
Added https://github.com/google/web-stories-wp/issues/8238 for enabling the feature flag + discussion.